File: solutions/0172-factorial-trailing-zeroes/factorial-trailing-zeroes.cpp
// Given an integer n, return the number of trailing zeroes in n!.
//
// Example 1:
//
//
// Input: 3
// Output: 0
// Explanation: 3! = 6, no trailing zero.
//
// Example 2:
//
//
// Input: 5
// Output: 1
// Explanation: 5! = 120, one trailing zero.
//
// Note: Your solution should be in logarithmic time complexity.
//


class Solution {
public:
    int trailingZeroes(int n) {
        int ans = 0;
        for(int i=1; ; ++i) {
            long tmp = pow(5, i);
            if(tmp > n) break;
            ans += n/tmp;
        }

        return ans;
    }
    
    long pow5(int n) {
        if(n == 1) return 5;
        return pow5(n-1) * 5;
    }
};

File: solutions/0172-factorial-trailing-zeroes/factorial-trailing-zeroes.py
# Given an integer n, return the number of trailing zeroes in n!.
#
# Example 1:
#
#
# Input: 3
# Output: 0
# Explanation: 3! = 6, no trailing zero.
#
# Example 2:
#
#
# Input: 5
# Output: 1
# Explanation: 5! = 120, one trailing zero.
#
# Note: Your solution should be in logarithmic time complexity.
#


class Solution:
    def trailingZeroes(self, n: int) -> int:
        count = 0
        while n >= 5:
            n = n // 5
            count += n
        return count


# 递归
# class Solution:
#     def trailingZeroes(self, n: int) -> int:
#         if n == 0: return 0
#         return n // 5 + self.trailingZeroes(n // 5)
